File: Codewars/Two_sum/two_sum.js
// Two Sum
// Codewars problem link https://www.codewars.com/kata/52c31f8e6605bcc646000082

// Example:
// target number = 15
// array = [12, 5, 32, 6, 10, 7]
// returns [1, 4]
// 1 is the index of 5 and 4 is the index of 10, 10 + 5 = target number

function two_sum(numbers, n) {
    let values = []

    for(let i=0; i<numbers.length; i++) {
        // a + b = n
        // To find (a) subtract (n) to (b) and to find (b) subtract (n) to (a) 
        let a = n - numbers[i]

        // 8 = 5 + 3; 5 = 8 - 3; 3 = 8 - 5

        if(values.includes(a)) {
            let index = values.indexOf(a)
            return [index, i]
        } else {
            values.push(numbers[i])
        }
    }

    return 0
}

let test_1 = [12, 5, 32, 6, 10, 7]
let target_1 = 15

let test_2 = [1, 2, 3]
let target_2 = 4

console.log(two_sum(test_1, target_1))
console.log(two_sum(test_2, target_2))

File: Codewars/Two_sum/two_sum.py
# TWO SUM

def two_sum(numbers, n):
    # Run for loop then Store the number and the index of that number
    # indexes = {1: 0, 2: 1, 3: 2}
    indexes = {}

    for i, b in enumerate(numbers):
        # a + b = n
        # To find (a) subtract (n) to (b) and to find (b) subtract (n) to (a) 
        a = n - b

        # Kinda like 3 + 5 = 8; 3 = 8 - 5; 5 = 8 - 3 

        if a in indexes:
            # Returns index
            return [indexes[a], i]
        elif a not in indexes:
            indexes[b] = i

print(two_sum([1, 2, 3], 4))
print(two_sum([12, 5, 32, 6, 10, 7], 15))